### ISSUE TITLE
About plotting 3D loss landscapes

### ISSUE THREAD

**Issue by mountains-high** [hiyouga/AMP-Regularizer#5]
Hello, 

Thank you for your very impressive and fine work.
Could you share your code for plotting the loss_landscapes, please? 
I tried to use the original work of Hao Li, Zheng Xu, Gavin Taylor, Christoph Studer, and Tom Goldstein. [Visualizing the Loss Landscape of Neural Nets](https://arxiv.org/abs/1712.09913). NIPS, 2018.
But it was giving some errors and was not able to plot them.

I would appreciate your help
Thank you

**Comment by hiyouga** [comment]
Here is my script to generate 3D loss landscapes
```python
import os
import time
import torch
import models
import argparse
import numpy as np
import torch.nn as nn
import matplotlib.pyplot as plt
from data_utils import load_cv_data
from mpl_toolkits.mplot3d import Axes3D


def load_data(args):
    # return your dataloaders
    return dataloaders


def create_model(args, checkpoint):
    # return the pretrained model
    return model


def evaluate(model, dataloader, criterion, args):
    total_loss, n_sample = 0, 0
    with torch.no_grad():
        for i_batch, (inputs, targets) in enumerate(dataloader):
            inputs, targets = inputs.to(args.device), targets.to(args.device)
            outputs = model(inputs)
            loss = criterion(outputs, targets)
            total_loss += loss.item() * targets.size(0)
            n_sample += targets.size(0)
    return total_loss / n_sample


def normalize_directions_for_weights(directions, weights):
    assert len(directions) == len(weights)
    for direction, weight in zip(directions, weights):
        if direction.dim() <= 1:
            direction.fill_(0)
        else:
            for d, w in zip(direction, weight):
                d.mul_(w.norm() / (d.norm() + 1e-10))


def create_random_directions(model):
    weights = [param.data for param in model.parameters()]
    directions = [torch.randn(w.size(), dtype=w.dtype, device=w.device) for w in weights]
    normalize_directions_for_weights(directions, weights)
    return weights, directions


def update_weights1D(model, weights, directions, step):
    for p, w, d in zip(model.parameters(), weights, directions):
        p.data = w + step * d


def update_weights2D(model, weights, directions_x, directions_y, step_x, step_y):
    for p, w, d_x, d_y in zip(model.parameters(), weights, directions_x, directions_y):
        p.data = w + step_x * d_x + step_y * d_y


def draw1D(args):
    points = np.linspace(-1.5, 1.5, 61)
    train_losses = list()
    test_losses = list()
    total_num = len(points)
    train_dataloader, test_dataloader = load_data(args)
    model = create_model(args, args.checkpoint)
    criterion = nn.CrossEntropyLoss()
    weights, directions = create_random_directions(model)
    for i, point in enumerate(points):
        print(f"computing #{i+1}, total {total_num} points")
        update_weights1D(model, weights, directions, point)
        train_loss = evaluate(model, train_dataloader, criterion, args)
        test_loss = evaluate(model, test_dataloader, criterion, args)
        train_losses.append(train_loss)
        test_losses.append(test_loss)
    np.savetxt(f"{args.timestamp}_train_loss_landscape1D.csv", train_losses, fmt='%.6f', delimiter=None)
    np.savetxt(f"{args.timestamp}_test_loss_landscape1D.csv", test_losses, fmt='%.6f', delimiter=None)
    plt.figure(figsize=(6, 4))
    plt.plot(points, train_losses, label='Train Loss', color='#1f77b4', linestyle='--', linewidth=1)
    plt.plot(points, test_losses, label='Test Loss', color='#1f77b4', linestyle='-', linewidth=1)
    plt.legend()
    plt.title('Loss Landscape')
    plt.xlabel('perturbation')
    plt.ylabel('loss')
    plt.ylim(args.plot_lim)
    plt.savefig(f"{args.timestamp}_loss_landscape1D.pdf", format='pdf', dpi=900, bbox_inches='tight')


def draw2D(args):
    point_num = 31
    points_x = np.linspace(-1.2, 1.2, point_num)
    points_y = np.linspace(-1.2, 1.2, point_num)
    points_X, points_Y = np.meshgrid(points_x, points_y)
    train_losses = np.zeros((point_num, point_num))
    test_losses = np.zeros((point_num, point_num))
    total_num = point_num ** 2
    train_dataloader, test_dataloader = load_data(args)
    model = create_model(args, args.checkpoint)
    criterion = nn.CrossEntropyLoss()
    weights, directions_x = create_random_directions(model)
    weights, directions_y = create_random_directions(model)
    for i in range(point_num):
        for j in range(point_num):
            print(f"computing #{i*point_num+j+1}, total {total_num} points")
            update_weights2D(model, weights, directions_x, directions_y, points_X[i][j], points_Y[i][j])
            train_loss = evaluate(model, train_dataloader, criterion, args)
            test_loss = evaluate(model, test_dataloader, criterion, args)
            train_losses[i][j] = train_loss
            test_losses[i][j] = test_loss
    np.savetxt(f"{args.timestamp}_train_loss_landscape2D.csv", train_losses, fmt='%.6f', delimiter=',')
    np.savetxt(f"{args.timestamp}_test_loss_landscape2D.csv", test_losses, fmt='%.6f', delimiter=',')
    if args.plot_dim == 2:
        levels = np.linspace(args.plot_lim[0], args.plot_lim[1], 10)
        plt.figure(figsize=(6, 4))
        CS = plt.contour(points_X, points_Y, train_losses, levels=levels, linewidths=1.0)
        plt.clabel(CS, inline=1, fontsize=10)
        plt.savefig(f"{args.timestamp}_train_loss_landscape2D.pdf", format='pdf', dpi=900, bbox_inches='tight')
        plt.figure(figsize=(6, 4))
        CS = plt.contour(points_X, points_Y, test_losses, levels=levels, linewidths=1.0)
        plt.clabel(CS, inline=1, fontsize=10)
        plt.savefig(f"{args.timestamp}_test_loss_landscape2D.pdf", format='pdf', dpi=900, bbox_inches='tight')
    elif args.plot_dim == 3:
        fig = plt.figure(figsize=(6, 4))
        ax = Axes3D(fig)
        ax.plot_surface(points_X, points_Y, train_losses, rcount=point_num, ccount=point_num, cmap=plt.cm.coolwarm, antialiased=False)
        ax.set_zlim(args.plot_lim)
        plt.savefig(f"{args.timestamp}_train_loss_landscape3D.pdf", format='pdf', dpi=900, bbox_inches='tight')
        fig = plt.figure(figsize=(6, 4))
        ax = Axes3D(fig)
        ax.plot_surface(points_X, points_Y, test_losses, rcount=point_num, ccount=point_num, cmap=plt.cm.coolwarm, antialiased=False)
        ax.set_zlim(args.plot_lim)
        plt.savefig(f"{args.timestamp}_test_loss_landscape3D.pdf", format='pdf', dpi=900, bbox_inches='tight')


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Drawer', formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument('--batch_size', type=int, default=256)
    parser.add_argument('--plot_dim', type=int, default=1, help='dimension of figures')
    parser.add_argument('--plot_lim', type=float, nargs=2, default=(-0.5, 5.0), help='limits of loss values')
    parser.add_argument('--device', type=str, default=None, choices=['cpu', 'cuda'])
    args = parser.parse_args()
    args.num_classes = num_classes[args.dataset]
    args.timestamp = str(int(time.time()))
    args.device = torch.device(args.device) if args.device else torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    plt.switch_backend('Agg')
    if args.plot_dim == 1:
        draw1D(args)
    else:
        draw2D(args)
```

**Comment by mountains-high** [comment]
Thank you very much
